### PR TITLE
feat: NL Signal Intake (#52), Session Wiring (#56), and FedRAMP Security Documentation

### LINKED ISSUE #52: Natural language signal intake: user-initiated investigation via NL text input
## Summary

Enable operators and external agents to trigger Kubernaut investigations by describing a problem in natural language. The API Frontend triages the request using its own LLM and K8s tools, identifies affected resources, and creates RemediationRequest CRDs directly via the K8s API.

This provides a "manual investigation trigger" that doesn't require an existing alert. Operators can type "pods in payments namespace are crash-looping after the last deploy" and Kubernaut investigates, diagnoses, and remediates through the standard CRD lifecycle pipeline.

**Supersedes**: jordigilh/kubernaut#714 (originally designed as a KA extraction endpoint; moved to AF after architectural reassessment — see rationale below).

## Architectural Rationale

The original design (kubernaut#714) placed NL extraction in KA via a `POST /api/v1/signal/extract` endpoint, with AF as a thin proxy. That design assumed AF had no LLM.

After reassessment, AF requires its own LLM for orchestration (deciding which MCP tool to call) and has direct K8s API access. Since AF already has both capabilities, the NL → structured signal extraction can happen entirely within AF's agent loop:

- **AF's LLM** interprets the user's natural language and plans the triage
- **AF's K8s tools** query events, pods, and deployments to discover actual problems
- **AF creates RR CRDs directly** via K8s API (same pattern as Gateway)
- **No cross-service call to KA** is needed for extraction
- **No Gateway routing** is needed — AF writes CRDs directly

KA's 3 MCP tools (`kubernaut_investigate`, `kubernaut_enrich`, `kubernaut_select_workflow`) remain unchanged — they all operate on existing RRs and are not involved in RR creation.

## Flow

```
User (Console, MCP client, or A2A agent)
  |
  | "Investigate the problems in namespace payments"
  v
API Frontend (LLM + K8s tools)
  |
  | AF's LLM: "Open-ended query — need to triage"
  |
  | 1. get_events(namespace=payments)
  | 2. list_pods(namespace=payments)
  | 3. get_deployments(namespace=payments)
  |
  | AF's LLM correlates results:
  |   - payment-svc: CrashLoopBackOff (OOMKilled x12)
  |   - checkout-worker: ImagePullBackOff
  |
  | 4. create_rr(resource=deploy/payment-svc, alert_type=OOMKilled, ...)
  | 5. create_rr(resource=deploy/checkout-worker, alert_type=ImagePullBackOff, ...)
  v
K8s API (direct CRD creation)
  |
  v
Full CRD lifecycle
  (SP -> AA -> KA investigation -> WFE -> EA)
```

## Entry Points

| Consumer | Protocol into AF | AF then... |
|----------|-----------------|------------|
| Kubernaut Console | MCP tool call | Triages via K8s tools, creates RR(s) |
| External AI agent | A2A `SendMessage` | Triages via K8s tools, creates RR(s) |
| MCP client (IDE, Slack bot) | MCP tool call | Triages via K8s tools, creates RR(s) |

## Implementation Scope

| Item | Notes |
|------|-------|
| AF K8s tools for triage | `get_events`, `list_pods`, `get_deployments` (read-only K8s API queries) |
| AF `create_rr` tool | Direct CRD creation via controller-runtime client (same pattern as `pkg/gateway/k8s/client.go`) |
| AF LLM triage prompt | System prompt guiding the LLM to triage namespace-wide or resource-specific queries |
| Dedup check before creation | Query existing RRs for same resource before creating (resource-level dedup) |
| Console NL input UI | Text input component (part of #16 scope) |
| Unit + integration tests | Triage logic, CRD creation, dedup, end-to-end NL → RR |

## Dependencies

- #14 (API Frontend Service) — AF service boundary and hosting model
- #16 (Kubernaut Console) — provides the UI entry point

## Related

- jordigilh/kubernaut#714 (original issue, now closed — superseded by this)
- jordigilh/kubernaut#703 (KA MCP Interactive Mode)
- #14 (API Frontend Service)
- #15 (A2A Protocol Support)
- #16 (Kubernaut Console)
- PROPOSAL-EXT-001 section 3.2 (needs update to reflect AF-owned extraction)

### LINKED ISSUE #56: feat(api): InvestigationSession CRD — AF's A2A-to-K8s bridge for session persistence
## Summary

Define and implement the InvestigationSession CRD — AF's lightweight bookkeeping resource that links A2A tasks to the kubernaut remediation pipeline. This CRD enables session persistence across AF restarts and user reconnections.

## Context

- #54: A2A interaction lifecycle (CRD design resolution in design update comment)
- #52: NL signal intake
- kubernaut#703: KA MCP Interactive Mode

## Design

The InvestigationSession CRD is AF-only — KA and AA do not know about it.

### CRD Schema

```yaml
apiVersion: apifrontend.kubernaut.ai/v1alpha1
kind: InvestigationSession
metadata:
  name: inv-<uuid>
  namespace: kubernaut-system
  labels:
    kubernaut.ai/rr-name: <rr-name>
    kubernaut.ai/user: <username>
spec:
  # Reference to the RemediationRequest this session is investigating
  remediationRequestRef:
    name: <rr-name>
    namespace: <rr-namespace>
  # A2A task ID for client reconnection
  a2aTaskID: <task-id>
  # User identity from JWT
  userIdentity:
    username: <username>
    groups: [<group1>, <group2>]
  # How the user joined: 'start' (manual signal) or 'takeover' (inflight join)
  joinMode: start | takeover
status:
  # Current session phase
  phase: Active | Completed | Cancelled | Disconnected
  # Reference to the AIAnalysis CRD (populated after AA is created)
  aiAnalysisRef: <aa-name>
  # KA interactive session ID (from kubernaut_investigate response)
  kaSessionID: <session-id>
  # Timestamps
  startedAt: <time>
  completedAt: <time>
  # Disconnect tracking
  disconnectedAt: <time>
  reconnectedAt: <time>
```

### Lifecycle

1. **Created** by AF when user initiates investigation (after triage, before or after RR creation)
2. **Updated** by AF as the flow progresses:
   - `status.aiAnalysisRef` set when AA is discovered via CRD watch
   - `status.kaSessionID` set when AF connects to KA via MCP
   - `status.phase` transitions: Active → Completed/Cancelled
3. **Read** by AF on reconnect: query by user identity label to find active sessions
4. **Garbage collected**: TTL-based cleanup for completed/cancelled sessions

### Reconnection Flow

1. User reconnects (SSE reconnect or new A2A request)
2. AF reads InvestigationSession CRDs with label `kubernaut.ai/user=<username>` and `status.phase=Active`
3. AF reads the referenced RR/AA state from K8s API
4. AF presents current investigation status to user
5. If KA interactive session is still active: resume. If expired: inform user of current autonomous state.

### Multi-Session Support

Multiple InvestigationSession CRDs can exist for the same RR (different users, different sessions over time). The label `kubernaut.ai/rr-name` enables lookup by RR, while `kubernaut.ai/user` enables lookup by user.

## Implementation Tasks

- [ ] CRD type definitions (`api/apifrontend/v1alpha1/investigationsession_types.go`)
- [ ] CRD YAML generation (`make generate manifests`)
- [ ] InvestigationSession creation in AF's investigation flow
- [ ] InvestigationSession status updates at each phase
- [ ] Reconnection logic: read InvestigationSession → read RR/AA → present state
- [ ] TTL-based garbage collection for terminal sessions
- [ ] Unit tests for lifecycle management
- [ ] Helm chart: CRD YAML included in deployment

## Dependencies

- #1: Project scaffolding (controller-runtime, CRD infrastructure)
- #54: A2A interaction lifecycle (flow that creates/updates this CRD)

## Not In Scope

- KA does not read or write this CRD
- AA controller does not read or write this CRD
- This CRD is not part of the kubernaut pipeline (RR → SP → AA → WFE → EA)

### PR DESCRIPTION
**Pull request by jordigilh**
## Summary

- **NL Signal Intake (#52):** 6 new triage tools (`af_list_events`, `af_get_pods`, `af_get_workloads`, `af_resolve_owner`, `af_check_existing_rr`, `af_create_rr`) with per-request user impersonation for read-only K8s queries and singleflight deduplication for RR creation
- **Session Production Wiring (#56):** `SessionServiceDecorator` enriches sessions with authenticated user identity; `DynamicClientFactory` pattern for impersonated Kubernetes clients
- **Security Hardening (SEC-05, FED-04):** User impersonation enforces least-privilege on triage tools; structured `tool.invoked` audit events emitted for every tool execution (NIST AU-12)
- **FedRAMP Documentation:** 5 substantive security/architecture documents mapping directly to implemented code — Authentication & RBAC, Service Boundary, Security Testing Pipeline, Audit Event Catalog, and Tool Execution Model

### Commit 1: Core Feature (`feat(tools)`)
- 6 new tool implementations in `internal/tools/af_*.go`
- `internal/validate/` package for RFC 1123 input validation
- `internal/auth/dynamic_impersonation.go` — `DynamicClientFactory` for per-request impersonation
- `internal/session/decorator.go` — `SessionServiceDecorator` for session enrichment
- Updated `internal/agent/root.go` with RBAC guard, metrics, and audit callbacks
- Extended `rbac_roles.yaml` with new tool grants across 3 roles
- Updated ClusterRole with read permissions for core K8s resources

### Commit 2: Tests (`test`)
- 7 new test files for all `af_*` tools + session decorator
- Updated 5 existing test files for new tool count (14→20) and prefix acceptance

### Commit 3: Test Plans (`docs(test)`)
- IEEE 829 test plans for #52 and #56

### Commit 4: Security Documentation (`docs(security)`)
- `docs/security/AUDIT_EVENT_CATALOG.md` — All event types mapped to NIST controls
- `docs/security/AUTHENTICATION_AND_RBAC.md` — Auth flow, RBAC matrix, impersonation model
- `docs/security/SERVICE_BOUNDARY.md` — Trust boundary, ingress/egress, rate limiting
- `docs/security/SECURITY_TESTING.md` — Testing pyramid, CI gates, vulnerability scanning
- `docs/design/TOOL_EXECUTION_MODEL.md` — Callback chain, client resolution, output safety

## Test plan

- [ ] `make test-unit` passes with race detector (all 77 test files)
- [ ] `make lint` passes
- [ ] `make validate-maturity-ci` passes
- [ ] New tools registered in MCP registry and agent card
- [ ] RBAC role counts match expected values (sre:20, ai-orchestrator:16, observability:8)
- [ ] Impersonation factory creates user-scoped clients (verified in `af_*_test.go`)

Closes #52, closes #56

Made with [Cursor](https://cursor.com)

### CODE REVIEW

**Review by jordigilh — COMMENTED** [review]
## Comprehensive Readiness Audit — PR #89

### Scope
NL Signal Intake (#52), Session Wiring (#56), FedRAMP Security Documentation
+3868/-47 across 40 files (6 new tools, session decorator, impersonation factory, validation, docs)

---

### Summary by Persona

| Persona | Pass | Findings |
|---------|------|----------|
| **Prod Security** | 4/7 | SEC-1 (P1), SEC-2 (P2), SEC-3 (P2) |
| **Architecture** | 5/7 | ARCH-1 (P1), ARCH-2 (P2) |
| **SRE** | 4/5 | SRE-1 (P2) |
| **QE / Test Quality** | 3/7 | TQ-1 (P1), TQ-2 (P2), TQ-3 (P2), TQ-4 (P3) |
| **FedRAMP** | 5/6 | FED-1 (P3) |
| **Product** | 5/5 | — |
| **UX** | 5/5 | — |

---

### Coverage Assessment

New production code ~900 LoC across 8 files; test files provide ~1200 LoC.
- `internal/tools/af_*` — each has dedicated tests with happy + validation + concurrency. **Good.**
- `internal/session/decorator.go` — tested but assertions are **weak** (see TQ-2).
- `internal/auth/dynamic_impersonation.go` — **NO dedicated test file** (TQ-1).
- `internal/validate/k8s.go` — **NO dedicated test file** (TQ-1).

### Findings below (12 inline comments).

File: internal/auth/dynamic_impersonation.go [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+			return nil, fmt.Errorf("impersonation requires authenticated user identity in context")
+		}
+
+		impCfg, err := NewImpersonatedConfig(baseCfg, identity.Username, identity.Groups)
+		if err != nil {
+			return nil, fmt.Errorf("creating impersonated config: %w", err)
+		}
+
+		client, err := dynamic.NewForConfig(impCfg)
+		if err != nil {
```
Comment (jordigilh) [inline internal/auth/dynamic_impersonation.go:39]: **[SEC-1] P1 — ProdSec** | No circuit breaker on impersonated clients

`dynamic.NewForConfig(impCfg)` creates a new client per tool invocation without circuit breaker wrapping. The shared SA `k8sClient` uses `resilience.NewResilientDynamicClient` — impersonated clients bypass this.

Under API server pressure, triage tools will:
1. Not trip a breaker to protect the cluster
2. Not share retry/backoff logic
3. Create unbounded TCP connections (one per tool call)

**Suggested fix:** Wrap with the same CB transport, or create a pool of impersonated clients per user.

File: internal/auth/dynamic_impersonation.go [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+// user's identity. Read-only triage tools use this to enforce least-privilege:
+// the user only sees what their RBAC permits, not the AF ServiceAccount's broader view.
+type DynamicClientFactory func(ctx context.Context) (dynamic.Interface, error)
+
+// NewImpersonatingDynamicFactory returns a DynamicClientFactory that creates an
+// impersonated dynamic.Interface based on the UserIdentity in the context.
+// If no identity is present, it returns an error (fail-closed).
+// baseCfg is the AF ServiceAccount's rest.Config (never mutated).
+func NewImpersonatingDynamicFactory(baseCfg *rest.Config) DynamicClientFactory {
+	return func(ctx context.Context) (dynamic.Interface, error) {
```
Comment (jordigilh) [inline internal/auth/dynamic_impersonation.go:27]: **[TQ-1] P1 — QE** | No dedicated unit tests for `dynamic_impersonation.go`

`NewImpersonatingDynamicFactory` and `StaticDynamicFactory` have **zero direct test coverage**. Missing scenarios:
1. `nil` identity in context → error
2. Empty `Username` → error
3. `StaticDynamicFactory(nil)` behavior
4. Successful path with identity → client created

This is security-critical code (impersonation) that should have isolation tests independent of integration via tool tests.

File: internal/agent/root.go [hunk 3]
```diff
@@ -214,3 +235,79 @@ func FilterToolsByRoles(roles []string, allTools []tool.Tool) []tool.Tool {
...
+			}
+		}
+		return nil, nil
+	}
+
+	return before, after
+}
+
+// newAuditToolCallback returns an AfterToolCallback that emits a structured
+// audit event for every tool invocation (FedRAMP AU-12 compliance).
```
Comment (jordigilh) [inline internal/agent/root.go:298]: **[SEC-2] P2 — ProdSec** | Audit callback may leak raw error strings

`newAuditToolCallback` puts `toolErr.Error()` directly into audit event `Detail["error"]`. While K8s errors pass through `ToUserFriendlyError` at tool boundaries, non-K8s errors (JSON parse, context cancelled, etc.) may include internal paths or stack info.

**Suggested fix:** Apply `security.RedactError(toolErr)` before storing in Detail, consistent with other audit emission points.

File: internal/agent/root.go [hunk 4]
```diff
@@ -214,3 +235,79 @@ func FilterToolsByRoles(roles []string, allTools []tool.Tool) []tool.Tool {
 	}
 	return result
 }
+
+// newMetricsToolCallbacks returns Before/After callbacks that track tool call
+// metrics: af_tool_calls_total (counter) and af_tool_call_duration_seconds (histogram).
+// Safe for concurrent use via sync.Map keyed by FunctionCallID.
+func newMetricsToolCallbacks(toolCalls *prometheus.CounterVec, toolDuration *prometheus.HistogramVec) (llmagent.BeforeToolCallback, llmagent.AfterToolCallback) {
+	var starts sync.Map
+
```
Comment (jordigilh) [inline internal/agent/root.go:251]: **[SRE-1] P2 — SRE** | `sync.Map` entries may leak on panic/cancel

`newMetricsToolCallbacks` stores `time.Now()` keyed by `FunctionCallID` in `before`. If `after` never runs (goroutine panic, context cancellation before completion), the entry remains in the map forever.

For long-running services this is a slow memory leak.

**Suggested fix:** Either:
1. Use `LoadAndDelete` in `after` (already done), plus periodic sweep of entries older than a threshold, or
2. Document that leaked entries are bounded by LLM call rate (acceptable if volume is low)

File: internal/tools/af_create_rr.go [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+// maxDescriptionLen is the maximum length for RR description (truncated, not rejected).
+const maxDescriptionLen = 2048
+
+// CreateRRArgs defines the input for af_create_rr.
+type CreateRRArgs struct {
+	Namespace   string `json:"namespace"`
+	Kind        string `json:"kind"`
+	Name        string `json:"name"`
+	Severity    string `json:"severity,omitempty"`
+	Description string `json:"description"`
```
Comment (jordigilh) [inline internal/tools/af_create_rr.go:38]: **[SEC-3] P2 — ProdSec** | `severity` field is unvalidated free-form string

`args.Severity` is written directly into CRD spec and labels without validation against an allowlist. An attacker with tool access could inject arbitrary severity values that bypass alert routing logic or confuse dashboards.

**Suggested fix:** Validate against `["critical", "high", "medium", "low", "info"]` before creating the RR.

File: internal/tools/af_check_existing_rr.go [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+		return CheckExistingRRResult{}, fmt.Errorf("%w: kind must not be empty", ErrInvalidInput)
+	}
+	if args.Name == "" {
+		return CheckExistingRRResult{}, fmt.Errorf("%w: name must not be empty", ErrInvalidInput)
+	}
+
+	labelSel := fmt.Sprintf("kubernaut.ai/target-kind=%s,kubernaut.ai/target-name=%s", args.Kind, args.Name)
+	list, err := client.Resource(rrGVR).Namespace(args.Namespace).List(ctx, metav1.ListOptions{
+		LabelSelector: labelSel,
+	})
```
Comment (jordigilh) [inline internal/tools/af_check_existing_rr.go:56]: **[ARCH-1] P1 — Architecture** | Label selector values not validated

`kind` and `name` are interpolated into the Kubernetes `LabelSelector` without validation through `validate.LabelValue()`. Malformed values (containing `/`, `>63 chars`, unicode) could produce:
1. Invalid selectors that error at the API
2. Unexpected matching behavior

The `validate.LabelValue` function exists in this PR — use it:
```go
if err := validate.LabelValue(args.Kind); err != nil { ... }
if err := validate.LabelValue(args.Name); err != nil { ... }
```

File: internal/validate/k8s.go [hunk 7]
```diff
@@ -0,0 +1,41 @@
+// Package validate provides reusable input validation helpers for Kubernetes
+// resource names and namespaces, wrapping k8s.io/apimachinery/pkg/util/validation
```
Comment (jordigilh) [inline internal/validate/k8s.go:2]: **[TQ-1b] P1 — QE** | No unit test file for `internal/validate/k8s.go`

This package provides security-critical input validation (`Namespace`, `ResourceName`, `LabelValue`) used by all 6 new tools. It has **zero direct tests**.

Indirect coverage via tool tests is not sufficient — boundary cases (63-char limit, leading/trailing hyphens, unicode, empty string) must be explicitly verified.

**Required:** Create `internal/validate/k8s_test.go` with table-driven tests covering RFC 1123 and Kubernetes label validation edge cases.

File: internal/session/decorator.go [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+	}
+
+	cfg := &CreateConfig{
+		A2ATaskID: sc.TaskID,
+		UserIdentity: v1alpha1.SessionUser{
+			Username: identity.Username,
+			Groups:   identity.Groups,
+		},
+		RemediationRef: sc.RemediationRef,
+	}
```
Comment (jordigilh) [inline internal/session/decorator.go:72]: **[ARCH-2] P2 — Architecture** | `RemediationRef` never populated in current wiring

The `CreateContext` struct has `RemediationRef` but the only producer (`launcher.go` `buildBeforeExecuteCallback`) only sets `TaskID`. This means:
1. CRD labels `LabelRRName`/`LabelRRNamespace` will always be empty
2. `RemediationRequestRef` in CRD spec will be zero-value

This is either dead code (remove) or a forward-compat placeholder (document with TODO).

File: internal/session/decorator.go [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+// then delegates to the inner service.
+func (d *ServiceDecorator) Create(ctx context.Context, req *adksession.CreateRequest) (*adksession.CreateResponse, error) {
+	sc := CreateContextFromContext(ctx)
+	if sc == nil {
+		return d.inner.Create(ctx, req)
+	}
+
+	identity := auth.UserIdentityFromContext(ctx)
+	if identity == nil || identity.Username == "" {
+		return nil, fmt.Errorf("session creation requires authenticated user identity")
```
Comment (jordigilh) [inline internal/session/decorator.go:62]: **[TQ-2] P2 — QE** | Decorator test doesn't assert enriched State contents

The test for "enriches State with CreateConfig" verifies `Create` succeeds but **never inspects** `req.State[StateKeyCreateConfig]` to confirm:
- `A2ATaskID` matches input
- `UserIdentity.Username` matches context identity
- `Groups` are propagated

Without value assertions, a regression that populates State with wrong values would pass silently.

File: internal/tools/af_create_rr.go [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+// CreateRRResult is the output of af_create_rr.
+type CreateRRResult struct {
+	RRID          string `json:"rr_id"`
+	Message       string `json:"message"`
+	AlreadyExists bool   `json:"already_exists,omitempty"`
+}
+
+// rrCreateGroup provides singleflight deduplication per fingerprint.
+var rrCreateGroup singleflight.Group
+
```
Comment (jordigilh) [inline internal/tools/af_create_rr.go:55]: **[ARCH-3] P3 — Architecture** | Process-wide `singleflight.Group` may interact across tests

The package-level `rrCreateGroup` is shared across all parallel tests and (in production) all concurrent requests. This is intentional for dedup but:
1. Test isolation: parallel tests with same fingerprint interact
2. Production: same fingerprint from **different users** shares one flight (winner's context runs the create)

Document that dedup is user-agnostic (acceptable for this use case) and note test isolation risk.

File: internal/tools/af_list_events.go [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+		item := &list.Items[i]
+		reason, _, _ := unstructured.NestedString(item.Object, "reason")
+		involvedKind, _, _ := unstructured.NestedString(item.Object, "involvedObject", "kind")
+
+		if args.Reason != "" && reason != args.Reason {
+			continue
+		}
+		if args.Kind != "" && involvedKind != args.Kind {
+			continue
+		}
```
Comment (jordigilh) [inline internal/tools/af_list_events.go:70]: **[TQ-3] P2 — QE** | `involved_kind` filter has no test coverage

Production code at this line filters events by `involvedObject.kind`, but `af_list_events_test.go` only tests the `reason` filter. The `involved_kind` path could regress without detection.

**Add:** A test case with events for different kinds, filtering to verify only matching kind is returned.

File: internal/tools/af_resolve_owner.go [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+		return empty, fmt.Errorf("%w: %v", ErrInvalidInput, err)
+	}
+	if err := validate.ResourceName(args.Name); err != nil {
+		return empty, fmt.Errorf("%w: %v", ErrInvalidInput, err)
+	}
+	if args.Kind == "" {
+		return empty, fmt.Errorf("%w: kind must not be empty", ErrInvalidInput)
+	}
+
+	currentKind, currentName := args.Kind, args.Name
```
Comment (jordigilh) [inline internal/tools/af_resolve_owner.go:115]: **[TQ-4] P3 — QE** | Owner chain tests don't cover mid-chain failures or cycle detection

The `visited` map and `maxOwnerDepth` (10 hops) logic have no dedicated test coverage:
1. Cycle (A→B→A) → should stop
2. Max depth reached → should return partial chain
3. Unsupported kind mid-chain → should break with partial result
4. Controller vs non-controller owner selection (`pickOwnerRef`)

These are correctness-critical for triage accuracy.

File: internal/auth/dynamic_impersonation.go (hunk 1 above)
Comment (jordigilh) [reply internal/auth/dynamic_impersonation.go:39]: Fixed in 6645476. `NewImpersonatingDynamicFactory` now accepts variadic `ClientWrapper` functions. `main.go` passes the shared `K8sCircuitBreaker` wrapper so every impersonated client is protected by the same CB instance used for the SA client.

File: internal/auth/dynamic_impersonation.go (hunk 2 above)
Comment (jordigilh) [reply internal/auth/dynamic_impersonation.go:27]: Fixed in 6645476. Added `internal/auth/dynamic_impersonation_test.go` with 6 test cases covering: nil identity, empty username, successful client creation, wrapper application order, nil static client, and non-nil static client.

File: internal/agent/root.go (hunk 3 above)
Comment (jordigilh) [reply internal/agent/root.go:298]: Fixed in 6645476. The audit callback now uses `security.RedactError(toolErr)` which strips URLs, file paths, JWTs, bearer tokens, and stack traces before writing to the audit log.

File: internal/agent/root.go (hunk 4 above)
Comment (jordigilh) [reply internal/agent/root.go:251]: Fixed in 6645476. Added periodic sweep goroutine (every 5 min) that evicts entries older than 5 minutes. Also documented the leak analysis in the function comment: worst case at 100 RPM with 100% loss is ~140KB/day — negligible for a long-running service.

File: internal/tools/af_create_rr.go (hunk 5 above)
Comment (jordigilh) [reply internal/tools/af_create_rr.go:38]: Fixed in 6645476. Added `validSeverities` map and validation check: severity must be one of `critical`, `high`, `medium`, `low`, `info` (or empty/omitted). Also added test cases UT-AF-052-059 (invalid rejected), UT-AF-052-060 (valid accepted), and UT-AF-052-061 (empty accepted).

File: internal/tools/af_check_existing_rr.go (hunk 6 above)
Comment (jordigilh) [reply internal/tools/af_check_existing_rr.go:56]: Fixed in 6645476. `kind` and `name` now validated via `validate.LabelValue()` after the empty checks. Added test cases UT-AF-052-048 (kind with slash rejected) and UT-AF-052-049 (name exceeding 63 chars rejected).

File: internal/validate/k8s.go (hunk 7 above)
Comment (jordigilh) [reply internal/validate/k8s.go:2]: Fixed in 6645476. Added `internal/validate/k8s_test.go` with Ginkgo table-driven boundary tests covering Namespace (10 cases), ResourceName (7 cases), and LabelValue (12 cases) including edge cases: max length boundaries, path traversal, unicode, and leading/trailing special chars.

File: internal/session/decorator.go (hunk 8 above)
Comment (jordigilh) [reply internal/session/decorator.go:72]: Fixed in 6645476. Added a TODO comment documenting `RemediationRef` as a forward-compatibility placeholder for v1.6 when the A2A protocol carries remediation context in task params (tracked in issue #54). Currently only `TaskID` is populated by the `BeforeExecuteCallback`.

File: internal/session/decorator.go (hunk 9 above)
Comment (jordigilh) [reply internal/session/decorator.go:62]: Fixed in 6645476. The decorator test now explicitly asserts `req.State[StateKeyCreateConfig]` is a `*CreateConfig` with the expected `A2ATaskID`, `UserIdentity.Username`, `UserIdentity.Groups`, `RemediationRef.Namespace`, and `RemediationRef.Name` values.

File: internal/tools/af_create_rr.go (hunk 10 above)
Comment (jordigilh) [reply internal/tools/af_create_rr.go:55]: Fixed in 6645476. Added documentation to the `rrCreateGroup` variable explaining that deduplication is intentionally user-agnostic: concurrent RR creation for the same target fingerprint is deduplicated regardless of initiator. The check_existing_rr safety net prevents duplicate CRDs regardless. Parallel tests sharing flights is by design.

File: internal/tools/af_list_events.go (hunk 11 above)
Comment (jordigilh) [reply internal/tools/af_list_events.go:70]: Fixed in 6645476. Added UT-AF-052-009 (filter by `involved_kind` only) and UT-AF-052-010 (combined filter by both `reason` and `involved_kind`) tests to `af_list_events_test.go`.

File: internal/tools/af_resolve_owner.go (hunk 12 above)
Comment (jordigilh) [reply internal/tools/af_resolve_owner.go:115]: Fixed in 6645476. Added 4 new tests to `af_resolve_owner_test.go`:
- UT-AF-052-039: Cycle detection (A->B->A) stops traversal with partial chain
- UT-AF-052-040: Unsupported kind mid-chain returns partial result (Pod->Job->CustomKind stops at Job)
- UT-AF-052-041: Max depth (10) stops traversal on deep chains
- UT-AF-052-042: Controller owner ref preferred over non-controller
